src/core/lib/event_engine/posix_engine/posix_endpoint.cc

```diff
@@ -0,0 +1,1135 @@
+  handle_->OrphanHandle(on_done_, nullptr, "");
+  delete on_read_;
+  delete on_write_;
+  delete on_error_;
+}
+
+PosixEndpointImpl::PosixEndpointImpl(EventHandle* handle,
+                                     PosixEngineClosure* on_done,
+                                     std::shared_ptr<EventEngine> engine,
+                                     const PosixTcpOptions& options)
+    : sock_(PosixSocketWrapper(handle->WrappedFd())),
+      on_done_(on_done),
+      traced_buffers_(),
+      handle_(handle),
+      poller_(handle->Poller()),
+      engine_(engine) {
+  PosixSocketWrapper sock(handle->WrappedFd());
+  fd_ = handle_->WrappedFd();
+  GPR_ASSERT(options.resource_quota != nullptr);
+  memory_owner_ = options.resource_quota->memory_quota()->CreateMemoryOwner(
+      *sock.PeerAddressString());
+  self_reservation_ = memory_owner_.MakeReservation(sizeof(PosixEndpointImpl));
+  local_address_ = *sock.LocalAddress();
+  peer_address_ = *sock.PeerAddress();
+  target_length_ = static_cast<double>(options.tcp_read_chunk_size);
+  bytes_read_this_round_ = 0;
+  min_read_chunk_size_ = options.tcp_min_read_chunk_size;
+  max_read_chunk_size_ = options.tcp_max_read_chunk_size;
+  tcp_zerocopy_send_ctx_ = absl::make_unique<TcpZerocopySendCtx>(
+      options.tcp_tx_zerocopy_max_simultaneous_sends,
+      options.tcp_tx_zerocopy_send_bytes_threshold);
+  frame_size_tuning_enabled_ = grpc_core::IsTcpFrameSizeTuningEnabled();
+  if (options.tcp_tx_zero_copy_enabled &&
+      !tcp_zerocopy_send_ctx_->MemoryLimited() && poller_->CanTrackErrors()) {
+#ifdef GRPC_LINUX_ERRQUEUE
+    const int enable = 1;
+    auto err =
+        setsockopt(fd_, SOL_SOCKET, SO_ZEROCOPY, &enable, sizeof(enable));
+    if (err == 0) {
+      tcp_zerocopy_send_ctx_->SetEnabled(true);
+    } else {
+      gpr_log(GPR_ERROR, "Failed to set zerocopy options on the socket.");
+    }
+#endif
+  }
+
+#ifdef GRPC_HAVE_TCP_INQ
+  int one = 1;
+  if (setsockopt(fd_, SOL_TCP, TCP_INQ, &one, sizeof(one)) == 0) {
+    inq_capable_ = true;
+  } else {
+    gpr_log(GPR_DEBUG, "cannot set inq fd=%d errno=%d", fd_, errno);
+    inq_capable_ = false;
+  }
+#else
+  inq_capable_ = false;
+#endif /* GRPC_HAVE_TCP_INQ */
+
+  on_read_ = PosixEngineClosure::ToPermanentClosure(
+      [this](absl::Status status) { HandleRead(std::move(status)); });
+  on_write_ = PosixEngineClosure::ToPermanentClosure(
+      [this](absl::Status status) { HandleWrite(std::move(status)); });
+  on_error_ = PosixEngineClosure::ToPermanentClosure(
+      [this](absl::Status status) { HandleError(std::move(status)); });
+
+  // Start being notified on errors if poller can track errors.
+  if (poller_->CanTrackErrors()) {
+    Ref();
+    handle_->NotifyOnError(on_error_);
+  }
+}
+
+std::unique_ptr<PosixEndpoint> CreatePosixEndpoint(
+    EventHandle* handle, PosixEngineClosure* on_shutdown,
+    std::shared_ptr<EventEngine> engine, const EndpointConfig& config) {
+  GPR_ASSERT(handle != nullptr);
```

Comment:
Handle is created by the poller and its ownership is transferred to the endpoint because it doesnt need to be kept around by the code which created it. I updated the comment in CreatePosixEndpoint in the header file to make this clear. So I think we can keep the handle as a pointer.